keps/sig-node/20190925-enable-seccomp-by-default.md

```diff
@@ -0,0 +1,335 @@
+of the targeted release**.
+
+For enhancements that make changes to code or processes/procedures in core Kubernetes i.e., [kubernetes/kubernetes], we require the following Release Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These checklist items _must_ be updated for the enhancement to be released.
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+A proposal to enhance the current seccomp support, enabling it by default in new deployments by leveraging a new audit-only feature.
+
+## Motivation
+
+Kubernetes supports Seccomp in some capacity since v1.3. But it is [disabled by default](https://github.com/kubernetes/kubernetes/issues/81115), as highlighted by the recent [security audit](https://github.com/kubernetes/kubernetes/issues/81146). 
+Enabling it poses a few challenges to users due to the absence of 1) auditing capabilities, 2) user-friendly ways to create and maintain profiles, and 3) mechanisms to synchronise custom profiles across nodes. 
+
+Removing such barriers of entry for seccomp usage in Kubernetes clusters, will encourage adoption and provide a safer baseline across the ecosystem.
+
+
+### Goals
+
+- Add audit-mode support for logging violations instead of blocking them.
+- Set clusters to use the new audit profile (i.e. `runtime/default-audit`) by default.  
```

Comment:
@tallclair yes, that is feasible. I can think of two ways to approach this, both involves amending the profile before sending it over to the CRI: 

1) *Audit only unspecified actions* by replacing the profile's defaultAction to `SCMP_ACT_LOG`. Very useful to keep allowed and blocked rules intact, but providing users a way to find extra syscalls that may need to be allowed.
2) *Audit all syscalls that are not whitelisted* by replacing all defined blocking actions and the defaultAction with `SCMP_ACT_LOG`. This answers questions such as "why my application is being blocked even though I am running it in audit mode?" 

The first approach sounds reasonable and would work fine with the default profile which currently does not blacklist anything. However, the second might be more user-friendly for custom profiles that may have blocking/tracing rules.

If we were to do this, I would go for the first approach and use a meaningful name for the flag inside the `SeccompOptions`. Maybe something in the lines of `AuditOnlyUnspecifiedSyscalls`?

Reply:
A caveat here is that this would not work with the current `runtime/default` profiles. This would only work for the new kubernetes and/or custom profiles, as we would be able to manipulate them.